### ISSUE TITLE
Singular value clamping code?

### ISSUE THREAD

**Issue by johntiger1** [ajbrock/BigGAN-PyTorch#57]
I am looking for the code that does the clamping of the singular values from the weight matrices:
![image](https://user-images.githubusercontent.com/14852840/77236563-1fbd7d80-6b96-11ea-9282-c25f31961e9f.png)

(i.e. page 6 of the arxiv paper)

but can't seem to find it in the training loop. Does anyone know where it is? 

Thanks

**Comment by ajbrock** [comment]
This is not included in BigGAN-PyTorch. As that section indicates, BigGAN collapse is not successfully mitigated by any one of a variety of normalization, clipping, or clamping strategies, even though they successfully treat the symptoms of spectral explosion. (Spectral Explosion is also probably an attack from a ghost anime). If you want to try SV clamping for your own experiments, all the tools are provided if you subclass an `SN` layer and implement eq (4).

**Comment by johntiger1** [comment]
Wow, thanks for the quick reply! Sure, but computing the singular values at every single update is the main challenge I would say. Was hoping to find the Arnoldi iteration method in Pytorch somewhere  

EDIT: seems like your code in https://github.com/ajbrock/BigGAN-PyTorch/blob/a5557079924c3070b39e67f2eaea3a52c0fb72ab/layers.py

is most of the way there! Thanks!
